cmd/jujud/agent/bootstrap.go

```diff
@@ -190,43 +256,54 @@ func (c *BootstrapCommand) Run(_ *cmd.Context) error {
-	}
-	addrs, err := instances[0].Addresses(callCtx)
-	if err != nil {
-		return errors.Annotate(err, "bootstrap instance addresses")
+	info, ok := agentConfig.StateServingInfo()
+	if !ok {
+		return fmt.Errorf("bootstrap machine config has no state serving info")
 	}
 
-	// When machine addresses are reported from state, they have
-	// duplicates removed.  We should do the same here so that
-	// there is not unnecessary churn in the mongo replicaset.
-	// TODO (cherylj) Add explicit unit tests for this - tracked
-	// by bug #1544158.
-	addrs = network.MergedAddresses([]network.Address{}, addrs)
+	addrs := []network.Address{network.NewAddress("127.0.0.1")}
+	if !isCAAS {
+		instanceLister, ok := env.(environs.InstanceLister)
+		if !ok {
+			// this should never happened.
+			return errors.NotValidf("InstanceLister missing for IAAS controller provider")
+		}
+		instances, err := instanceLister.Instances(callCtx, []instance.Id{args.BootstrapMachineInstanceId})
+		if err != nil {
+			return errors.Annotate(err, "getting bootstrap instance")
+		}
+		addrs, err = instances[0].Addresses(callCtx)
+		if err != nil {
+			return errors.Annotate(err, "bootstrap instance addresses")
+		}
 
-	// Generate a private SSH key for the controllers, and add
-	// the public key to the environment config. We'll add the
-	// private key to StateServingInfo below.
-	privateKey, publicKey, err := sshGenerateKey(config.JujuSystemKey)
-	if err != nil {
-		return errors.Annotate(err, "failed to generate system key")
-	}
-	authorizedKeys := config.ConcatAuthKeys(args.ControllerModelConfig.AuthorizedKeys(), publicKey)
-	newConfigAttrs[config.AuthorizedKeysKey] = authorizedKeys
+		// When machine addresses are reported from state, they have
+		// duplicates removed.  We should do the same here so that
+		// there is not unnecessary churn in the mongo replicaset.
+		// TODO (cherylj) Add explicit unit tests for this - tracked
+		// by bug #1544158.
+		addrs = network.MergedAddresses([]network.Address{}, addrs)
+
+		// Generate a private SSH key for the controllers, and add
+		// the public key to the environment config. We'll add the
+		// private key to StateServingInfo below.
+		privateKey, publicKey, err := sshGenerateKey(config.JujuSystemKey)
+		if err != nil {
+			return errors.Annotate(err, "failed to generate system key")
+		}
+		info.SystemIdentity = privateKey
 
-	// Generate a shared secret for the Mongo replica set, and write it out.
-	sharedSecret, err := mongo.GenerateSharedSecret()
-	if err != nil {
-		return err
-	}
-	info, ok := agentConfig.StateServingInfo()
-	if !ok {
-		return fmt.Errorf("bootstrap machine config has no state serving info")
+		authorizedKeys := config.ConcatAuthKeys(args.ControllerModelConfig.AuthorizedKeys(), publicKey)
+		newConfigAttrs[config.AuthorizedKeysKey] = authorizedKeys
+
+		// Generate a shared secret for the Mongo replica set, and write it out.
+		sharedSecret, err := mongo.GenerateSharedSecret()
```

Comment:
don't we need to do this also for k8s controllers?

Reply:
no, we don't.